File: typescript/grade-school/grade-school.ts
interface Roster {
  [key: string]: string[]
}

export class GradeSchool {
  _roster: Roster = {}

  roster(): Roster {
    return JSON.parse(JSON.stringify(this._roster))
  }

  add(name: string, level: number): void {
    this._roster[level] ??= []
    const studentsForLevel = this._roster[level]
    this.remove(name)

    this._roster[level] = [...studentsForLevel, name].sort()
  }

  grade(level: number): string[] {
    return this.roster()[level] ?? []
  }

  private remove(name: string): void {
    for(const level in this._roster) {
      this._roster[level] = this._roster[level].filter(student => {
        return student !== name
      })
    }
  }
}

File: ruby/grade-school/grade_school.rb
class School
  attr_reader :grades

  def initialize
    @grades = (1..7).to_h { |i| [i, []] }
  end

  def add(student, grade)
    return false if roster.include?(student)

    @grades[grade] << student
    @grades[grade].sort!
    true
  end

  def roster
    grades.values.flatten
  end

  def grade(year)
    grades[year]
  end
end
